File: src/main/java/org/goafabric/core/data/logic/PractitionerLogic.java
package org.goafabric.core.data.logic;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.transaction.Transactional;
import org.goafabric.core.data.controller.vo.Practitioner;
import org.goafabric.core.data.logic.mapper.PractitionerMapper;
import org.goafabric.core.data.repository.PractitionerRepository;

import java.util.List;

@ApplicationScoped
@Transactional
public class PractitionerLogic {
    private final PractitionerMapper practitionerMapper;

    private final PractitionerRepository practitionerRepository;

    public PractitionerLogic(PractitionerMapper practitionerMapper, PractitionerRepository practitionerRepository) {
        this.practitionerMapper = practitionerMapper;
        this.practitionerRepository = practitionerRepository;
    }

    public Practitioner getById(String id) {
        return practitionerMapper.map(
                practitionerRepository.findById(id));
    }

    public void deleteById(String id) {
        practitionerRepository.deleteById(id);
    }

    public List<Practitioner> findByGivenName(String givenName) {
        return practitionerMapper.map(
                practitionerRepository.findByGivenNameStartsWithIgnoreCase(givenName));
    }

    public List<Practitioner> findByFamilyName(String familyName) {
        return practitionerMapper.map(
                practitionerRepository.findByFamilyNameStartsWithIgnoreCase(familyName));
    }

    public Practitioner save(Practitioner practitioner) {
        return practitionerMapper.map(practitionerRepository.save(
                practitionerMapper.map(practitioner)));
    }

}

File: src/main/kotlin/org/goafabric/core/organization/logic/PractitionerLogic.kt
package org.goafabric.core.organization.logic

import jakarta.enterprise.context.ApplicationScoped
import jakarta.transaction.Transactional
import org.goafabric.core.organization.controller.dto.Practitioner
import org.goafabric.core.organization.logic.mapper.PractitionerMapper
import org.goafabric.core.organization.persistence.PractitionerRepository

@ApplicationScoped
@Transactional
class PractitionerLogic(
    private val mapper: PractitionerMapper,
    private val repository: PractitionerRepository
) {
    fun getById(id: String): Practitioner = mapper.map(repository.findById(id) ?: throw IllegalArgumentException("Practitioner not found: $id"))

    fun deleteById(id: String) = repository.deleteById(id)

    fun findByGivenName(givenName: String): List<Practitioner> =
        mapper.map(repository.findByGivenNameStartsWith(givenName))

    fun findByFamilyName(familyName: String): List<Practitioner> =
        mapper.map(repository.findByFamilyNameStartsWith(familyName))

    fun save(practitioner: Practitioner): Practitioner =
        mapper.map(repository.save(mapper.map(practitioner)))
}
